h2o-core/src/main/java/water/FailedNodeWatchdogExtension.java

```diff
@@ -0,0 +1,155 @@
+                return new_args;
+            }
+        }
+        return args;
+    }
+
+
+    public void validateArguments() {
+        if (watchdogClientRetryTimeout < 0) {
+            H2O.parseFailed("Watchdog client retry timeout has to be positive: " + watchdogClientRetryTimeout);
+        }
+    }
+
+    @Override
+    public void onLocalNodeStarted() {
+        new FailedNodeWatchdogThread().start();
+        H2O.SELF._heartbeat._watchdog_client = watchDogClient;
+    }
+
+
+    /**
+     * This method checks whether the client is disconnected from this node due to some problem such as client or network
+     * is unreachable.
+     */
+    private static void handleClientDisconnect(H2ONode node) {
+        if(node._heartbeat._watchdog_client){
+            Log.warn("Watchdog client " + node + " disconnected!");
+            BullyClientGoneTask tsk = new BullyClientGoneTask(node);
+            Log.warn("Asking the rest of the nodes in the cloud whether watchdog client is really gone.");
+            if(((BullyClientGoneTask)tsk.doAllNodes()).consensus) {
+                Log.fatal("Stopping H2O cloud since the watchdog client is disconnected from all nodes in the cluster!");
+                H2O.shutdown(0);
+            }
+        }else if(node._heartbeat._client) {
+            Log.warn("Client "+ node +" disconnected!");
+        }
+
+        // in both cases remove the client
+        if(node._heartbeat._client){
+            H2O.removeClient(node);
+            if(H2O.isFlatfileEnabled()){
+                H2O.removeNodeFromFlatfile(node);
+            }
+        }
+    }
+
+    /**
+     * Helper MR task used to detect consensus on the fact that bully client is disconnected from the network
+     */
+    private static class BullyClientGoneTask extends MRTask {
+        public boolean consensus = true;
+        private H2ONode clientNode;
+
+        public BullyClientGoneTask(H2ONode clientNode) {
+            this.clientNode = clientNode;
+        }
+
+        @Override
+        public void reduce(MRTask mrt) {
+            if(((BullyClientGoneTask)mrt).consensus) { // don't change the value in case negative consensus
+                for(H2ONode node: H2O.getClients()){
+                    // find the same client node on the other nodes
+                    if(node.equals(clientNode) && node._connection_closed){
+                        consensus = true;
+                        break;
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Thread used to run disconnect hooks on nodes who disconnects from the cloud
+     */
+    private static class FailedNodeWatchdogThread extends Thread {
+        final private int sleepMillis = 6000; // 6 seconds
+
+        public FailedNodeWatchdogThread() {
+            super("FailedNodeWatchdogThread");
+            setDaemon(true);
+        }
+
+        @Override
+        public void run() {
+            while (true) {
+
```

Comment:
Yup, the previous solution https://github.com/h2oai/h2o-3/commit/9851f14d112d96ac90d79982b2661cbf932581fa contained all this code in the HeartBeat, however we decided to make this optional feature which may be turned off.